File: app/auth/invites.py
"""Issuing and redeeming invitations.

The flow, and why it is shaped this way:

    admin names an email + role   ->  issue_invite()  -> raw token, once
    admin hands the link over     ->  (Slack, in person; never emailed)
    invitee opens the link        ->  validate_invite()
    invitee sets THEIR OWN password -> accept_invite() -> role attached

No password is ever generated for the invitee, shown to the admin, or derived
from the role. A role-derived password would be identical for everyone holding
that role and guessable by anyone who has seen the pattern, including people who
have left — and it would make "approved by X" unprovable, which defeats the
point of auditing at all.
"""

import hashlib
import secrets
import uuid
from datetime import UTC, datetime, timedelta

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.auth.service import bump_permissions_version
from app.db.models.authz import Role, RoleName, UserInvite, UserRole

# Long enough that guessing is hopeless; the token is the only thing standing
# between a stranger and a provisioned account.
_TOKEN_BYTES = 32

DEFAULT_INVITE_TTL = timedelta(hours=72)


class InviteError(Exception):
    """An invite cannot be used. The message is safe to show the invitee."""


def _hash_token(raw_token: str) -> str:
    """SHA-256 of the raw token.

    A plain hash, not a password KDF, is correct here: the token is 32 bytes of
    CSPRNG output, so there is no dictionary to attack and nothing for bcrypt's
    work factor to defend against.
    """
    return hashlib.sha256(raw_token.encode()).hexdigest()


def issue_invite(
    db: Session,
    *,
    email: str,
    role: RoleName,
    created_by: str | None,
    ttl: timedelta = DEFAULT_INVITE_TTL,
) -> tuple[UserInvite, str]:
    """Create an invite. Returns the row and the raw token, which is shown once.

    Re-inviting the same email is allowed and simply supersedes: any earlier
    unused invite is revoked, so a link that was shared to the wrong person
    stops working the moment a new one is issued.
    """
    normalised = email.strip().lower()

    role_row = db.scalar(select(Role).where(Role.name == role.value))
    if role_row is None:
        raise InviteError(f"Role {role.value} does not exist. Has the seed been run?")

    now = datetime.now(UTC)
    for stale in db.scalars(
        select(UserInvite).where(
            UserInvite.email == normalised,
            UserInvite.accepted_at.is_(None),
            UserInvite.revoked_at.is_(None),
        )
    ):
        stale.revoked_at = now

    raw_token = secrets.token_urlsafe(_TOKEN_BYTES)
    invite = UserInvite(
        email=normalised,
        role_id=role_row.id,
        token_hash=_hash_token(raw_token),
        expires_at=now + ttl,
        created_by=created_by,
    )
    db.add(invite)
    db.flush()
    return invite, raw_token


def validate_invite(db: Session, raw_token: str) -> UserInvite:
    """Return the invite this token unlocks, or raise InviteError.

    Every rejection says the same thing. Distinguishing "expired" from "already
    used" from "never existed" would confirm to a stranger which tokens are real.
    """
    invite = db.scalar(select(UserInvite).where(UserInvite.token_hash == _hash_token(raw_token)))

    if invite is None or invite.revoked_at is not None or invite.accepted_at is not None:
        raise InviteError("This invitation is no longer valid. Ask an admin for a new one.")

    # Stored tz-aware, but SQLite hands back naive datetimes; normalise so the
    # comparison cannot raise instead of rejecting.
    expires_at = invite.expires_at
    if expires_at.tzinfo is None:
        expires_at = expires_at.replace(tzinfo=UTC)

    if expires_at <= datetime.now(UTC):
        raise InviteError("This invitation is no longer valid. Ask an admin for a new one.")

    return invite


def accept_invite(db: Session, *, raw_token: str, user_id: str) -> UserInvite:
    """Redeem an invite for a freshly created user and attach the role.

    Called after the frontend has created the Better Auth user, so `user_id` is
    a real id. Marks the invite used in the same transaction as the grant: a
    half-applied invite would either burn a link with no role, or leave a link
    live after it had already been used.
    """
    invite = validate_invite(db, raw_token)

    invite.accepted_at = datetime.now(UTC)
    invite.accepted_user_id = user_id

    already_granted = db.scalar(
        select(UserRole).where(UserRole.user_id == user_id, UserRole.role_id == invite.role_id)
    )
    if already_granted is None:
        db.add(UserRole(user_id=user_id, role_id=invite.role_id, assigned_by=invite.created_by))

    # Without this the grant is invisible to any token already minted for them.
    bump_permissions_version(db, user_id)

    db.flush()
    return invite


def revoke_invite(db: Session, invite_id: uuid.UUID) -> None:
    """Cancel an unused invite, so its link stops working."""
    invite = db.get(UserInvite, invite_id)
    if invite is None or invite.accepted_at is not None:
        raise InviteError("No pending invitation to revoke.")
    invite.revoked_at = datetime.now(UTC)
    db.flush()

File: internal/auth/invites.go
package auth

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"

	"github.com/Optiminastic/tensor-core/internal/db/gen"
)

// InviteError is a message safe to show the invitee. Every rejection reason
// shares one message so a stranger cannot tell an expired token from a used one
// from one that never existed.
type InviteError struct{ Msg string }

func (e InviteError) Error() string { return e.Msg }

const (
	inviteTokenBytes = 32
	// DefaultInviteTTL is how long a fresh invite stays valid.
	DefaultInviteTTL = 72 * time.Hour

	invalidInviteMessage = "This invitation is no longer valid. Ask an admin for a new one."
	noPendingInviteMsg   = "No pending invitation to revoke."
)

func hashToken(raw string) string {
	sum := sha256.Sum256([]byte(raw))
	return hex.EncodeToString(sum[:])
}

func newRawToken() (string, error) {
	b := make([]byte, inviteTokenBytes)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(b), nil
}

// IssueInvite creates an invite for an email and role, superseding any pending
// invite for that email. It returns the stored row and the one-time raw token
// (never persisted). Run inside a transaction so the revoke and insert are
// atomic.
func IssueInvite(
	ctx context.Context, q *gen.Queries, email string, role RoleName, createdBy string, ttl time.Duration,
) (gen.UserInvite, string, error) {
	normalised := strings.ToLower(strings.TrimSpace(email))

	roleID, err := q.GetRoleIDByName(ctx, string(role))
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return gen.UserInvite{}, "", InviteError{
				Msg: fmt.Sprintf("Role %s does not exist. Has the seed been run?", role),
			}
		}
		return gen.UserInvite{}, "", err
	}

	if err := q.RevokePendingInvitesForEmail(ctx, normalised); err != nil {
		return gen.UserInvite{}, "", err
	}

	raw, err := newRawToken()
	if err != nil {
		return gen.UserInvite{}, "", err
	}

	createdByPtr := &createdBy
	invite, err := q.InsertInvite(ctx, gen.InsertInviteParams{
		ID:        uuid.New(),
		Email:     normalised,
		RoleID:    roleID,
		TokenHash: hashToken(raw),
		ExpiresAt: pgtype.Timestamptz{Time: time.Now().UTC().Add(ttl), Valid: true},
		CreatedBy: createdByPtr,
	})
	if err != nil {
		return gen.UserInvite{}, "", err
	}
	return invite, raw, nil
}

// ValidateInvite looks up a live invite by its raw token. Missing, revoked,
// accepted and expired all yield the same InviteError message.
func ValidateInvite(ctx context.Context, q *gen.Queries, rawToken string) (gen.UserInvite, error) {
	invite, err := q.GetInviteByTokenHash(ctx, hashToken(rawToken))
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return gen.UserInvite{}, InviteError{Msg: invalidInviteMessage}
		}
		return gen.UserInvite{}, err
	}
	if invite.RevokedAt.Valid || invite.AcceptedAt.Valid {
		return gen.UserInvite{}, InviteError{Msg: invalidInviteMessage}
	}
	if !time.Now().UTC().Before(invite.ExpiresAt.Time) {
		return gen.UserInvite{}, InviteError{Msg: invalidInviteMessage}
	}
	return invite, nil
}

// AcceptInvite redeems an invite for a newly created user and attaches the role,
// bumping the user's permissions version. It is one-shot: a used or expired
// invite is rejected. Run inside a transaction.
func AcceptInvite(ctx context.Context, q *gen.Queries, rawToken, userID string) (gen.UserInvite, error) {
	invite, err := ValidateInvite(ctx, q, rawToken)
	if err != nil {
		return gen.UserInvite{}, err
	}
	if err := q.MarkInviteAccepted(ctx, gen.MarkInviteAcceptedParams{
		ID: invite.ID, AcceptedUserID: &userID,
	}); err != nil {
		return gen.UserInvite{}, err
	}
	if err := q.InsertUserRole(ctx, gen.InsertUserRoleParams{
		UserID: userID, RoleID: invite.RoleID, AssignedBy: invite.CreatedBy,
	}); err != nil {
		return gen.UserInvite{}, err
	}
	if _, err := q.BumpPermissionsVersion(ctx, userID); err != nil {
		return gen.UserInvite{}, err
	}
	return invite, nil
}

// RevokeInvite kills a pending invite. A missing or already-accepted invite is
// an error; anything else is revoked.
func RevokeInvite(ctx context.Context, q *gen.Queries, inviteID uuid.UUID) error {
	invite, err := q.GetInviteByID(ctx, inviteID)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return InviteError{Msg: noPendingInviteMsg}
		}
		return err
	}
	if invite.AcceptedAt.Valid {
		return InviteError{Msg: noPendingInviteMsg}
	}
	return q.MarkInviteRevoked(ctx, inviteID)
}
